internal/app/machined/pkg/controllers/siderolink/userspace.go

```diff
@@ -0,0 +1,200 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package siderolink
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"sync/atomic"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"github.com/siderolabs/siderolink/pkg/wgtunnel"
+	"github.com/siderolabs/siderolink/pkg/wgtunnel/wgbind"
+	"github.com/siderolabs/siderolink/pkg/wgtunnel/wggrpc"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/siderolabs/talos/internal/pkg/ctxutil"
+	"github.com/siderolabs/talos/internal/pkg/endpoint"
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	"github.com/siderolabs/talos/pkg/machinery/resources/siderolink"
+)
+
+// UserspaceWireguardController imlements a controller that manages a Wireguard over GRPC tunnel in userspace.
+type UserspaceWireguardController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *UserspaceWireguardController) Name() string {
+	return "siderolink.UserspaceWireguardController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *UserspaceWireguardController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      siderolink.TunnelType,
+			ID:        optional.Some(siderolink.TunnelID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *UserspaceWireguardController) Outputs() []controller.Output {
+	return []controller.Output{}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *UserspaceWireguardController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	eg, ctx := errgroup.WithContext(ctx)
+
+	logger.WithOptions()
+
+	var (
+		tunnelDevice atomic.Pointer[wgtunnel.TunnelDevice]
+		tunnelRelay  atomic.Pointer[tunnelProps]
+	)
+
+	defer func() {
+		tunnelRelay.Load().Relay().Close()
```

Comment:
how does it work if the pointer is nil?

Reply:
All the types in the chain support being called on a nil pointer. That was the whole idea.